### ISSUE TITLE
Unexpected exceptions in duplicate stream may not be notified to subscribers

### ISSUE THREAD

**Issue by anuraaga** [line/armeria#1039]
I noticed that a refcnt exception in the duplicator stream gets logged but is not returned to the subscriber. This means that an exception causes the stream to just hang forever, with futures never completing. This might be ok, since the alternative is way too many `catch (Throwable)`, and it just means we have to be careful in righting code that can't throw exceptions, but want to hear what you think.

This sort of stacktrace gets logged when the stream gets hung.

```
io.netty.util.IllegalReferenceCountException: refCnt: 0 at io.netty.buffer.AbstractByteBuf.ensureAccessible(AbstractByteBuf.java:1415) 
[netty-buffer-4.1.21.Final.jar:4.1.21.Final] at io.netty.buffer.AbstractByteBuf.checkIndex(AbstractByteBuf.java:1354) 
[netty-buffer-4.1.21.Final.jar:4.1.21.Final] at io.netty.buffer.UnsafeByteBufUtil.getBytes(UnsafeByteBufUtil.java:481) 
[netty-buffer-4.1.21.Final.jar:4.1.21.Final] at io.netty.buffer.PooledUnsafeDirectByteBuf.getBytes(PooledUnsafeDirectByteBuf.java:131) 
[netty-buffer-4.1.21.Final.jar:4.1.21.Final] at io.netty.buffer.AbstractUnpooledSlicedByteBuf.getBytes(AbstractUnpooledSlicedByteBuf.java:243) 
[netty-buffer-4.1.21.Final.jar:4.1.21.Final] at io.netty.buffer.CompositeByteBuf.getBytes(CompositeByteBuf.java:862) 
[netty-buffer-4.1.21.Final.jar:4.1.21.Final] at io.netty.buffer.CompositeByteBuf.getBytes(CompositeByteBuf.java:44) 
[netty-buffer-4.1.21.Final.jar:4.1.21.Final] at io.netty.buffer.AbstractUnpooledSlicedByteBuf.getBytes(AbstractUnpooledSlicedByteBuf.java:243) 
[netty-buffer-4.1.21.Final.jar:4.1.21.Final] at io.netty.buffer.UnpooledHeapByteBuf.setBytes(UnpooledHeapByteBuf.java:262) 
[netty-buffer-4.1.21.Final.jar:4.1.21.Final] at io.netty.buffer.AbstractByteBuf.writeBytes(AbstractByteBuf.java:1080) 
[netty-buffer-4.1.21.Final.jar:4.1.21.Final] at io.netty.buffer.Unpooled.copiedBuffer(Unpooled.java:436) 
[netty-buffer-4.1.21.Final.jar:4.1.21.Final] at com.linecorp.armeria.common.stream.AbstractStreamMessageDuplicator$DownstreamSubscription.copy(AbstractStreamMessageDuplicator.java:766) 
[armeria-0.58.1.jar:?] at com.linecorp.armeria.common.stream.AbstractStreamMessageDuplicator$DownstreamSubscription.doSignalSingle(AbstractStreamMessageDuplicator.java:711) 
[armeria-0.58.1.jar:?] at com.linecorp.armeria.common.stream.AbstractStreamMessageDuplicator$DownstreamSubscription.doSignal(AbstractStreamMessageDuplicator.java:649) 
[armeria-0.58.1.jar:?] at com.linecorp.armeria.common.stream.AbstractStreamMessageDuplicator$DownstreamSubscription.signal(AbstractStreamMessageDuplicator.java:641) 
[armeria-0.58.1.jar:?] at com.linecorp.armeria.common.stream.AbstractStreamMessageDuplicator$DownstreamSubscription.request(AbstractStreamMessageDuplicator.java:624) 
[armeria-0.58.1.jar:?] at com.linecorp.armeria.client.HttpRequestSubscriber.operationComplete(HttpRequestSubscriber.java:95) 
[armeria-0.58.1.jar:?] at com.linecorp.armeria.client.HttpRequestSubscriber.operationComplete(HttpRequestSubscriber.java:48) 
[armeria-0.58.1.jar:?] at io.netty.util.concurrent.DefaultPromise.notifyListener0(DefaultPromise.java:512) [netty-common-4.1.21.Final.jar:4.1.21.Final] at io.netty.util.concurrent.DefaultPromise.notifyListeners0(DefaultPromise.java:505) [netty-common-4.1.21.Final.jar:4.1.21.Final] at io.netty.util.concurrent.DefaultPromise.notifyListenersNow(DefaultPromise.java:484) [netty-common-4.1.21.Final.jar:4.1.21.Final] at io.netty.util.concurrent.DefaultPromise.notifyListeners(DefaultPromise.java:425) [netty-common-4.1.21.Final.jar:4.1.21.Final] at io.netty.util.concurrent.DefaultPromise.trySuccess(DefaultPromise.java:104) [netty-common-4.1.21.Final.jar:4.1.21.Final] at io.netty.handler.codec.http2.Http2CodecUtil$SimpleChanne
```

**Comment by minwoox** [comment]
Oops~ I think we need to return the exception to the subscriber. I never realized that it can throw an exception there :)

**Comment by anuraaga** [comment]
It only can throw an exception due to programming error in armeria itself, not user code, so it's not so clear to me that we need to - not sure ;)

**Comment by anuraaga** [comment]
Well - I guess working with a pooled buffer may always have issues due to user error as well, so leaning towards checking the refcnt or exception after thinking a bit more.

**Comment by minwoox** [comment]
Closed by #1065
